## 1. Use pandas, or the BeautifulSoup package, or any other way you are comfortable with to transform the data in the table on the Wikipedia page into the above pandas dataframe.

In [27]:
from bs4 import BeautifulSoup
import requests
from geopy.geocoders import Nominatim
import pandas as pd
import numpy as np
import matplotlib.cm as cm
import matplotlib.colors as colors
from sklearn.cluster import KMeans
import folium

In [9]:
# download url data from internet
url = "https://en.wikipedia.org/wiki/List_of_postal_codes_of_Canada:_M"
source = requests.get(url).text
soup = BeautifulSoup(source, 'lxml')

In [10]:
# creat a new Dataframe
column_names = ['Postalcode','Borough','Neighborhood']
toronto = pd.DataFrame(columns = column_names)

In [11]:
table_contents=[]
table=soup.find('table')
for row in table.findAll('td'):
    cell = {}
    if row.span.text=='Not assigned':
        pass
    else:
        cell['PostalCode'] = row.p.text[:3]
        cell['Borough'] = (row.span.text).split('(')[0]
        cell['Neighborhood'] = (((((row.span.text).split('(')[1]).strip(')')).replace(' /',',')).replace(')',' ')).strip(' ')
        table_contents.append(cell)

df=pd.DataFrame(table_contents)
df['Borough']=df['Borough'].replace({'Downtown TorontoStn A PO Boxes25 The Esplanade':'Downtown Toronto Stn A',
                                             'East TorontoBusiness reply mail Processing Centre969 Eastern':'East Toronto Business',
                                             'EtobicokeNorthwest':'Etobicoke Northwest','East YorkEast Toronto':'East York/East Toronto',
                                             'MississaugaCanada Post Gateway Processing Centre':'Mississauga'})
df.head()

PostalCode           Borough                      Neighborhood
0        M3A        North York                         Parkwoods
1        M4A        North York                  Victoria Village
2        M5A  Downtown Toronto         Regent Park, Harbourfront
3        M6A        North York  Lawrence Manor, Lawrence Heights
4        M7A      Queen's Park     Ontario Provincial Government

In [12]:
print(df.shape)

(103, 3)


##### The dataframe has 103 Postal codes.

### 2. Now that you have built a dataframe of the postal code of each neighborhood along with the borough name and neighborhood name, in order to utilize the Foursquare location data, we need to get the latitude and the longitude coordinates of each neighborhood. 

In [13]:
def get_geocode(postal_code):
    # initialize your variable to None
    lat_lng_coords = None
    while(lat_lng_coords is None):
        g = geocoder.google('{}, Toronto, Ontario'.format(postal_code))
        lat_lng_coords = g.latlng
    latitude = lat_lng_coords[0]
    longitude = lat_lng_coords[1]
    return latitude,longitude

In [14]:
geo_df=pd.read_csv('https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DS0701EN-SkillsNetwork/labs_v1/Geospatial_Coordinates.csv')

In [15]:
geo_df.rename(columns={'Postal Code':'PostalCode'},inplace=True)
geo_merged = pd.merge(geo_df, df, on='PostalCode')

In [16]:
geo_data=geo_merged[['PostalCode','Borough','Neighborhood','Latitude','Longitude']]
geo_data.head()

PostalCode      Borough                            Neighborhood   Latitude  \
0        M1B  Scarborough                          Malvern, Rouge  43.806686   
1        M1C  Scarborough  Rouge Hill, Port Union, Highland Creek  43.784535   
2        M1E  Scarborough       Guildwood, Morningside, West Hill  43.763573   
3        M1G  Scarborough                                  Woburn  43.770992   
4        M1H  Scarborough                               Cedarbrae  43.773136   

   Longitude  
0 -79.194353  
1 -79.160497  
2 -79.188711  
3 -79.216917  
4 -79.239476

In [17]:
toronto_data=geo_data[geo_data['Borough'].str.contains("Toronto")]
toronto_data.head()

PostalCode                 Borough                    Neighborhood  \
37        M4E            East Toronto                     The Beaches   
40        M4J  East York/East Toronto              The Danforth  East   
41        M4K            East Toronto    The Danforth West, Riverdale   
42        M4L            East Toronto  India Bazaar, The Beaches West   
43        M4M            East Toronto                 Studio District   

     Latitude  Longitude  
37  43.676357 -79.293031  
40  43.685347 -79.338106  
41  43.679557 -79.352188  
42  43.668999 -79.315572  
43  43.659526 -79.340923

# Explore and cluster the neighborhoods in Toronto.

In [18]:
CLIENT_ID = 'SJQOUGR0LPTEZKP0CL5BTYIZOV2NMLOCV2UNWAE2COUPFBPL' # your Foursquare ID
CLIENT_SECRET = 'BJT2UTATRNTGGO3FAVIZQ21NNBO2SSTXONMZEI5BGOEQMJSG' # your Foursquare Secret
VERSION = '20210823'

In [19]:
def getNearbyVenues(names, latitudes, longitudes):
    radius=500
    LIMIT=100
    venues_list=[]
    for name, lat, lng in zip(names, latitudes, longitudes):
        print(name)
            
        # create the API request URL
        url = 'https://api.foursquare.com/v2/venues/explore?&client_id={}&client_secret={}&v={}&ll={},{}&radius={}&limit={}'.format(
            CLIENT_ID, 
            CLIENT_SECRET, 
            VERSION, 
            lat, 
            lng, 
            radius, 
            LIMIT)
            
        # make the GET request
        results = requests.get(url).json()["response"]['groups'][0]['items']
        
        # return only relevant information for each nearby venue
        venues_list.append([(
            name, 
            lat, 
            lng, 
            v['venue']['name'], 
            v['venue']['location']['lat'], 
            v['venue']['location']['lng'],  
            v['venue']['categories'][0]['name']) for v in results])
        nearby_venues = pd.DataFrame([item for venue_list in venues_list for item in venue_list])
    nearby_venues.columns = ['Neighborhood', 
                  'Neighborhood Latitude', 
                  'Neighborhood Longitude', 
                  'Venue', 
                  'Venue Latitude', 
                  'Venue Longitude', 
                  'Venue Category']
    
    return(nearby_venues)

In [20]:
toronto_venues = getNearbyVenues(names=toronto_data['Neighborhood'],
                                   latitudes=toronto_data['Latitude'],
                                   longitudes=toronto_data['Longitude']
                                  )

The Beaches
The Danforth  East
The Danforth West, Riverdale
India Bazaar, The Beaches West
Studio District
Lawrence Park
Davisville North
North Toronto West
Davisville
Moore Park, Summerhill East
Summerhill West, Rathnelly, South Hill, Forest Hill SE, Deer Park
Rosedale
St. James Town, Cabbagetown
Church and Wellesley
Regent Park, Harbourfront
Garden District, Ryerson
St. James Town
Berczy Park
Central Bay Street
Richmond, Adelaide, King
Harbourfront East, Union Station, Toronto Islands
Toronto Dominion Centre, Design Exchange
Commerce Court, Victoria Hotel
Roselawn
Forest Hill North & West
The Annex, North Midtown, Yorkville
University of Toronto, Harbord
Kensington Market, Chinatown, Grange Park
CN Tower, King and Spadina, Railway Lands, Harbourfront West, Bathurst Quay, South Niagara, Island airport
Enclave of M5E
First Canadian Place, Underground city
Christie
Dufferin, Dovercourt Village
Little Portugal, Trinity
Brockton, Parkdale Village, Exhibition Place
High Park, The Junction 

In [21]:
toronto_venues.head()

Neighborhood  Neighborhood Latitude  Neighborhood Longitude  \
0  The Beaches              43.676357              -79.293031   
1  The Beaches              43.676357              -79.293031   
2  The Beaches              43.676357              -79.293031   
3  The Beaches              43.676357              -79.293031   
4  The Beaches              43.676357              -79.293031   

                                Venue  Venue Latitude  Venue Longitude  \
0                   Glen Manor Ravine       43.676821       -79.293942   
1  The Big Carrot Natural Food Market       43.678879       -79.297734   
2                 Grover Pub and Grub       43.679181       -79.297215   
3                       Upper Beaches       43.680563       -79.292869   
4                 Seaspray Restaurant       43.678888       -79.298167   

      Venue Category  
0              Trail  
1  Health Food Store  
2                Pub  
3       Neighborhood  
4   Asian Restaurant

In [22]:
toronto_venues.groupby('Neighborhood').count()

Neighborhood Latitude  \
Neighborhood                                                                
Berczy Park                                                            60   
Brockton, Parkdale Village, Exhibition Place                           24   
CN Tower, King and Spadina, Railway Lands, Harb...                     15   
Central Bay Street                                                     62   
Christie                                                               17   
Church and Wellesley                                                   78   
Commerce Court, Victoria Hotel                                        100   
Davisville                                                             34   
Davisville North                                                        9   
Dufferin, Dovercourt Village                                           16   
Enclave of M4L                                                         16   
Enclave of M5E                                                         98   
First Canadian Place, Underground city                                100   
Forest Hill North & West                                                4   
Garden District, Ryerson                                              100   
Harbourfront East, Union Station, Toronto Islands                     100   
High Park, The Junction South                                          26   
India Bazaar, The Beaches West                                         20   
Kensington Market, Chinatown, Grange Park                              69   
Lawrence Park                                                           4   
Little Portugal, Trinity                                               44   
Moore Park, Summerhill East                                             1   
North Toronto West                                                     17   
Parkdale, Roncesvalles                                                 14   
Regent Park, Harbourfront                                              44   
Richmond, Adelaide, King                                              100   
Rosedale                                                                4   
Roselawn                                                                3   
Runnymede, Swansea                                                     34   
St. James Town                                                         91   
St. James Town, Cabbagetown                                            48   
Studio District                                                        37   
Summerhill West, Rathnelly, South Hill, Forest ...                     14   
The Annex, North Midtown, Yorkville                                    23   
The Beaches                                                             5   
The Danforth  East                                                      2   
The Danforth West, Riverdale                                           41   
Toronto Dominion Centre, Design Exchange                              100   
University of Toronto, Harbord                                         33   

                                                    Neighborhood Longitude  \
Neighborhood                                                                 
Berczy Park                                                             60   
Brockton, Parkdale Village, Exhibition Place                            24   
CN Tower, King and Spadina, Railway Lands, Harb...                      15   
Central Bay Street                                                      62   
Christie                                                                17   
Church and Wellesley                                                    78   
Commerce Court, Victoria Hotel                                         100   
Davisville                                                              34   
Davisville North                                                         9   
Dufferin, Dovercourt Village        

In [23]:
# one hot encoding
toronto_onehot = pd.get_dummies(toronto_venues[['Venue Category']], prefix="", prefix_sep="")
toronto_onehot.drop(['Neighborhood'],axis=1,inplace=True) 
toronto_onehot.insert(loc=0, column='Neighborhood', value=toronto_venues['Neighborhood'] )
toronto_onehot.shape

(1607, 240)

In [24]:
toronto_grouped = toronto_onehot.groupby('Neighborhood').mean().reset_index()
toronto_grouped.head()

Neighborhood  Afghan Restaurant  \
0                                        Berczy Park                0.0   
1       Brockton, Parkdale Village, Exhibition Place                0.0   
2  CN Tower, King and Spadina, Railway Lands, Har...                0.0   
3                                 Central Bay Street                0.0   
4                                           Christie                0.0   

    Airport  Airport Food Court  Airport Lounge  Airport Service  \
0  0.000000            0.000000        0.000000         0.000000   
1  0.000000            0.000000        0.000000         0.000000   
2  0.066667            0.066667        0.066667         0.133333   
3  0.000000            0.000000        0.000000         0.000000   
4  0.000000            0.000000        0.000000         0.000000   

   Airport Terminal  American Restaurant  Antique Shop  Aquarium  ...  \
0          0.000000                  0.0           0.0       0.0  ...   
1          0.000000                  0.0           0.0       0.0  ...   
2          0.133333                  0.0           0.0       0.0  ...   
3          0.000000                  0.0           0.0       0.0  ...   
4          0.000000                  0.0           0.0       0.0  ...   

   Thrift / Vintage Store  Tibetan Restaurant  Toy / Game Store  Trail  \
0                     0.0                 0.0               0.0    0.0   
1                     0.0                 0.0               0.0    0.0   
2                     0.0                 0.0               0.0    0.0   
3                     0.0                 0.0               0.0    0.0   
4                     0.0                 0.0               0.0    0.0   

   Train Station  Vegetarian / Vegan Restaurant  Video Game Store  \
0            0.0                       0.016667               0.0   
1            0.0                       0.000000               0.0   
2            0.0                       0.000000               0.0   
3            0.0                       0.016129               0.0   
4            0.0                       0.000000               0.0   

   Vietnamese Restaurant  Wine Bar  Yoga Studio  
0                    0.0       0.0     0.000000  
1                    0.0       0.0     0.041667  
2                    0.0       0.0     0.000000  
3                    0.0       0.0     0.016129  
4                    0.0       0.0     0.000000  

[5 rows x 240 columns]

In [25]:
def return_most_common_venues(row, num_top_venues):
    row_categories = row.iloc[1:]
    row_categories_sorted = row_categories.sort_values(ascending=False)
    
    return row_categories_sorted.index.values[0:num_top_venues]

In [28]:
num_top_venues = 10

indicators = ['st', 'nd', 'rd']

# create columns according to number of top venues
columns = ['Neighborhood']
for ind in np.arange(num_top_venues):
    try:
        columns.append('{}{} Most Common Venue'.format(ind+1, indicators[ind]))
    except:
        columns.append('{}th Most Common Venue'.format(ind+1))

# create a new dataframe
neighborhoods_venues_sorted = pd.DataFrame(columns=columns)
neighborhoods_venues_sorted['Neighborhood'] = toronto_grouped['Neighborhood']

for ind in np.arange(toronto_grouped.shape[0]):
    neighborhoods_venues_sorted.iloc[ind, 1:] = return_most_common_venues(toronto_grouped.iloc[ind, :], num_top_venues)

neighborhoods_venues_sorted.head()

Neighborhood 1st Most Common Venue  \
0                                        Berczy Park           Coffee Shop   
1       Brockton, Parkdale Village, Exhibition Place                  Café   
2  CN Tower, King and Spadina, Railway Lands, Har...       Airport Service   
3                                 Central Bay Street           Coffee Shop   
4                                           Christie         Grocery Store   

  2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
0                Bakery          Cocktail Bar    Seafood Restaurant   
1        Breakfast Spot           Coffee Shop         Grocery Store   
2      Airport Terminal                 Plane           Coffee Shop   
3                  Café        Sandwich Place                   Spa   
4                  Café                  Park            Restaurant   

  5th Most Common Venue 6th Most Common Venue 7th Most Common Venue  \
0            Restaurant              Beer Bar        Farmers Market   
1                   Gym         Burrito Place          Climbing Gym   
2    Airport Food Court        Airport Lounge       Harbor / Marina   
3      Department Store       Bubble Tea Shop          Burger Joint   
4                  Bank           Candy Store             Nightclub   

  8th Most Common Venue   9th Most Common Venue 10th Most Common Venue  
0           Cheese Shop          Clothing Store         Sandwich Place  
1     Convenience Store  Furniture / Home Store           Intersection  
2                   Bar        Sculpture Garden                Airport  
3                  Bank      Italian Restaurant             Restaurant  
4    Italian Restaurant      Athletics & Sports             Baby Store

Making Clusters for Neighborhood

In [29]:
# set number of clusters
kclusters = 5

toronto_grouped_clustering = toronto_grouped.drop('Neighborhood', 1)

# run k-means clustering
kmeans = KMeans(n_clusters=kclusters, random_state=0).fit(toronto_grouped_clustering)

# check cluster labels generated for each row in the dataframe
kmeans.labels_[0:10]

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1], dtype=int32)

In [30]:
# add clustering labels
neighborhoods_venues_sorted.insert(0, 'Cluster Labels', kmeans.labels_)

toronto_merged = toronto_data

# merge toronto_grouped with toronto_data to add latitude/longitude for each neighborhood
toronto_merged = toronto_merged.join(neighborhoods_venues_sorted.set_index('Neighborhood'), on='Neighborhood')

toronto_merged.head()

PostalCode                 Borough                    Neighborhood  \
37        M4E            East Toronto                     The Beaches   
40        M4J  East York/East Toronto              The Danforth  East   
41        M4K            East Toronto    The Danforth West, Riverdale   
42        M4L            East Toronto  India Bazaar, The Beaches West   
43        M4M            East Toronto                 Studio District   

     Latitude  Longitude  Cluster Labels 1st Most Common Venue  \
37  43.676357 -79.293031               1                   Pub   
40  43.685347 -79.338106               0                  Park   
41  43.679557 -79.352188               1      Greek Restaurant   
42  43.668999 -79.315572               1        Sandwich Place   
43  43.659526 -79.340923               1           Coffee Shop   

   2nd Most Common Venue 3rd Most Common Venue 4th Most Common Venue  \
37     Health Food Store      Asian Restaurant                 Trail   
40     Convenience Store     Afghan Restaurant                Museum   
41           Coffee Shop    Italian Restaurant        Ice Cream Shop   
42     Fish & Chips Shop      Sushi Restaurant                  Park   
43                  Café               Brewery   American Restaurant   

        5th Most Common Venue   6th Most Common Venue  \
37  Middle Eastern Restaurant     Moroccan Restaurant   
40                     Market     Martial Arts School   
41                       Café  Furniture / Home Store   
42                Coffee Shop                     Pub   
43                  Gastropub                  Bakery   

       7th Most Common Venue            8th Most Common Venue  \
37       Monument / Landmark  Molecular Gastronomy Restaurant   
40  Mediterranean Restaurant                      Men's Store   
41             Grocery Store                  Bubble Tea Shop   
42                Restaurant                    Movie Theater   
43               Yoga Studio                              Bar   

         9th Most Common Venue     10th Most Common Venue  
37  Modern European Restaurant         Miscellaneous Shop  
40          Mexican Restaurant  Middle Eastern Restaurant  
41        Caribbean Restaurant             Cosmetics Shop  
42               Burrito Place                    Brewery  
43                     Gay Bar  Latin American Restaurant

In [31]:
neighborhoods_venues_sorted.head()

Cluster Labels                                       Neighborhood  \
0               1                                        Berczy Park   
1               1       Brockton, Parkdale Village, Exhibition Place   
2               1  CN Tower, King and Spadina, Railway Lands, Har...   
3               1                                 Central Bay Street   
4               1                                           Christie   

  1st Most Common Venue 2nd Most Common Venue 3rd Most Common Venue  \
0           Coffee Shop                Bakery          Cocktail Bar   
1                  Café        Breakfast Spot           Coffee Shop   
2       Airport Service      Airport Terminal                 Plane   
3           Coffee Shop                  Café        Sandwich Place   
4         Grocery Store                  Café                  Park   

  4th Most Common Venue 5th Most Common Venue 6th Most Common Venue  \
0    Seafood Restaurant            Restaurant              Beer Bar   
1         Grocery Store                   Gym         Burrito Place   
2           Coffee Shop    Airport Food Court        Airport Lounge   
3                   Spa      Department Store       Bubble Tea Shop   
4            Restaurant                  Bank           Candy Store   

  7th Most Common Venue 8th Most Common Venue   9th Most Common Venue  \
0        Farmers Market           Cheese Shop          Clothing Store   
1          Climbing Gym     Convenience Store  Furniture / Home Store   
2       Harbor / Marina                   Bar        Sculpture Garden   
3          Burger Joint                  Bank      Italian Restaurant   
4             Nightclub    Italian Restaurant      Athletics & Sports   

  10th Most Common Venue  
0         Sandwich Place  
1           Intersection  
2                Airport  
3             Restaurant  
4             Baby Store

In [32]:
address = 'Toronto, CA'

geolocator = Nominatim(user_agent="ny_explorer")
location = geolocator.geocode(address)
latitude = location.latitude
longitude = location.longitude
print('The geograpical coordinate of Manhattan are {}, {}.'.format(latitude, longitude))

The geograpical coordinate of Manhattan are 43.6534817, -79.3839347.


###### Creating Map

In [33]:
map_clusters = folium.Map(location=[latitude, longitude], zoom_start=11)

# set color scheme for the clusters
x = np.arange(kclusters)
ys = [i + x + (i*x)**2 for i in range(kclusters)]
colors_array = cm.rainbow(np.linspace(0, 1, len(ys)))
rainbow = [colors.rgb2hex(i) for i in colors_array]

# add markers to the map
markers_colors = []
for lat, lon, poi, cluster in zip(toronto_merged['Latitude'], toronto_merged['Longitude'], toronto_merged['Neighborhood'], toronto_merged['Cluster Labels']):
    label = folium.Popup(str(poi) + ' Cluster ' + str(cluster), parse_html=True)
    folium.CircleMarker(
        [lat, lon],
        radius=5,
        popup=label,
        color=rainbow[cluster-1],
        fill=True,
        fill_color=rainbow[cluster-1],
        fill_opacity=0.7).add_to(map_clusters)
       
map_clusters